### DISCUSSION TITLE
How does Smart-Wallet work?

### DISCUSSION THREAD

**Discussion by anilhelvaci** [Agoric/agoric-sdk#6934] [Q&A] [answered]
Hello 👋 

I am trying to understand how the `smart-wallet` is working. Here's a list of questions I've gathered so far;

* In [wallet/README.md](https://github.com/Agoric/agoric-sdk/tree/master/packages/wallet#agoric-wallet), there's this phrase: 
  > The smart wallet has not yet subsumed all the features of the off-chain wallet. Until it does, we require the api directory.

  Does that mean features in the [Wallet API](https://docs.agoric.com/reference/wallet-api.html#wallet-api-commands) and [Wallet Bridge API](https://docs.agoric.com/reference/wallet-api.html#walletbridge-api-commands) are not supported?

* I assume `smart-wallet` is users' agent to the Mainnet.  If yes, how can we deploy contracts to the Mainnet using `smart-wallet`? If not, how?
* When I inspect the [Smart Wallet UI](https://github.com/Agoric/wallet-app) code;  there's this method that catches my attention [unserializeOfferFromStorage](https://github.com/Agoric/wallet-app/blob/1deefaa90b9bcb90dea3e9e0aa994197a400b0aa/src/service/Offers.ts#L44-L162). As far as I can understand, it listens for new offers written to the `chainStorage` and  wraps it in a way that [Smart Wallet Contract](https://github.com/Agoric/agoric-sdk/blob/bcb2ff8340dea145b3b07b56cf21d0d49f8b8f48/packages/smart-wallet/src/smartWallet.js#L526-L530) can understand and execute the offer. Two questions here;
  * Who writes the `offer` to the `chainStorage`?
    * If it's the `addOffer` returned to the dapp, like in the [dapp psm here](https://github.com/Agoric/dapp-psm/blob/db7cb7a808d2e016cab1aed9f6eefffced466275/src/components/WalletBridge.tsx#L108), does that mean the `addOffer` dapp-psm is using is not the same as in the [lib-wallet](https://github.com/Agoric/agoric-sdk/blob/bcb2ff8340dea145b3b07b56cf21d0d49f8b8f48/packages/wallet/api/src/lib-wallet.js#L1194-L1236)?
  * How's the connection between `Smart Wallet UI` and `Smart Wallet Contract` established?

Tagging you here @dckc like you said in the office hours.

Thanks 🙏

**Comment by dckc** [comment]
>   Does that mean features in the [Wallet API](https://docs.agoric.com/reference/wallet-api.html#wallet-api-commands) and [Wallet Bridge API](https://docs.agoric.com/reference/wallet-api.html#walletbridge-api-commands) are not supported?

That API is based on general-purpose object messaging between an off-chain persistent vats (such as in `ag-solo`) and on-chain vats (which are all persistent). This was implemented as so-called "mailbox" access: [MsgDeliverInbound](https://github.com/Agoric/agoric-sdk/blob/master/golang/cosmos/proto/agoric/swingset/msgs.proto#L13) protobuf messages to the chain were routed to the swingset kernel `mailbox` device, and in the other direction, replies were stored in a way that off-chain clients could get them with cosmos RPC. For mainnet-1 at least, this general-purpose object messaging is constrained to devnet:

 - https://github.com/Agoric/agoric-sdk/issues/5965

> * I assume `smart-wallet` is users' agent to the Mainnet.  If yes, how can we deploy contracts to the Mainnet using `smart-wallet`?

The smart wallet is not directly relevant to deployment at this time.

 1. bundle the contract using `bundle-source --cache-json ...`; note the resulting hash
 2. Put the bundle on chain using `agoric publish`, i.e. send a signed [InstallBundle](https://github.com/Agoric/agoric-sdk/blob/master/golang/cosmos/proto/agoric/swingset/msgs.proto#L11) protobuf message and use cosmos RPC and the hash to confirm that the bundle was installed without error
 3. call `E(zoe).installBundleID('b1-...hash...')` to get an installation
 4. call [E(zoe).startInstance(installation, ...)](https://docs.agoric.com/reference/zoe-api/zoe.html#e-zoe-startinstance-installation-issuerkeywordrecord-terms-privateargs)

The first 2 steps are permissionless: `bundle-source` is a local operation and `InstallBundle` is available to anyone willing to pay (note upcoming price increase: https://github.com/Agoric/agoric-sdk/issues/6858 ).

The other 2 steps involve executing JS on-chain:
  - On **devnet**, you can use `ag-solo` mailbox access as usual. For details, see [Dec 14 notes](https://github.com/Agoric/agoric-sdk/issues/6454#issuecomment-1351949397).
  - On **mainnet**, we have not yet delegated to anything less than the whole BLDer DAO, so use a [swingset.CoreEvalProposal](https://github.com/Agoric/agoric-sdk/blob/master/golang/cosmos/proto/agoric/swingset/swingset.proto#L12) as in [Discussion #6839](https://github.com/Agoric/agoric-sdk/discussions/6839).

It's possible that the BLDer DAO will delegate the `installBundleID` and `startInstance` steps to a contract governed by something like the economic committee. In that case, the smart wallet would be a way to make offers to this governed contract, perhaps using extra offer args for the args to `startInstance`. I'm not sure how we would support passing `privateArgs` without disclosing them. hm.

_more on the other questions separately..._

**Comment by dckc** [comment]
> * When I inspect the [Smart Wallet UI](https://github.com/Agoric/wallet-app) code;  there's this method that catches my attention [unserializeOfferFromStorage](https://github.com/Agoric/wallet-app/blob/1deefaa90b9bcb90dea3e9e0aa994197a400b0aa/src/service/Offers.ts#L44-L162). As far as I can understand, it listens for new offers written to the `chainStorage`

No, in this case "storage" refers to browser `localStorage` that the wallet uses to persist offers across page reloads and such as the user composes them.

https://github.com/Agoric/wallet-app/blob/1deefaa90b9bcb90dea3e9e0aa994197a400b0aa/src/util/storage.ts#L20-L23

>   * How's the connection between `Smart Wallet UI` and `Smart Wallet Contract` established?

via `MsgWalletSpendAction` protobuf messages.

https://github.com/Agoric/agoric-sdk/blob/7cd4bcfc6e730a269d6cea61cb8f6b3ebc927371/golang/cosmos/proto/agoric/swingset/msgs.proto#L16-L17

### Clients Compose, Sign, and Broadcast Smart Wallet Offers

The smart wallet UI helps the user compose such messages and then hands off to keplr to sign and broadcast:

https://github.com/Agoric/wallet-app/blob/1deefaa90b9bcb90dea3e9e0aa994197a400b0aa/src/util/keyManagement.ts#L507-L520

Now what goes in that `spendAction` string? Basically a serialized offer, using board IDs to preserve identity of brands and such.

It's perhaps easier to study CLI code.

In [agoric-cli/test/agops-perf-smoketest.sh](https://github.com/Agoric/agoric-sdk/blob/master/packages/agoric-cli/test/agops-perf-smoketest.sh) we have:

```
bin/agops psm swap --wantMinted 0.01 --feePct 0.01 --pair IST.USDC_axl >|"$OFFER"
```

The code to make such an offer is...

https://github.com/Agoric/agoric-sdk/blob/7cd4bcfc6e730a269d6cea61cb8f6b3ebc927371/packages/agoric-cli/src/lib/psm.js

```js
/**
 * @param {Instance} instance
 * @param {Record<string, Brand>} brands
 * @param {{ offerId: number, feePct?: number } &
 *         ({ wantMinted: number | undefined, giveMinted: number | undefined })} opts
 * @returns {BridgeAction}
 */
export const makePSMSpendAction = (instance, brands, opts) => {
  const method = opts.wantMinted
    ? 'makeWantMintedInvitation'
    : 'makeGiveMintedInvitation'; // ref psm.js
  const proposal = makePSMProposal(
    brands,
    opts,
    opts.feePct ? opts.feePct / 100 : undefined,
    // @ts-expect-error please update types. Not sure where pair goees.
    opts.pair[1],
  );

  console.warn('psm spend give', proposal.give);
  console.warn('psm spend want', proposal.want);

  // NB: not really a Proposal because the brands are not remotes
  // Instead they're copyRecord like  "{"boardId":"board0257","iface":"Alleged: IST brand"}" to pass through the boardId
  // mustMatch(harden(proposal), ProposalShape);

  /** @type {OfferSpec} */
  const offer = {
    id: opts.offerId,
    invitationSpec: {
      source: 'contract',
      instance,
      publicInvitationMaker: method,
    },
    proposal,
  };

  /** @type {BridgeAction} */
  const spendAction = {
    method: 'executeOffer',
    offer,
  };
  return harden(spendAction);
};
```

where `makePSMProposal` does...

```js
export const makePSMProposal = (brands, opts, fee = 0, anchor = 'AUSD') => {
...
  return {
    give: {
      In: { brand: brand.in, value: adjusted.in },
    },
    want: {
      Out: { brand: brand.out, value: adjusted.out },
    },
  };
```

Then `agoric wallet send --from "$WALLET" --offer "$OFFER"` is used to sign and broadcast (by shelling out to `agd`). `agops perf satisfaction ...` is a wrapper that waits for the disposition of the offer by polling `published.wallet.agoric1...` for offer status updates.

### Smart Wallet Contract executes offers from `MsgWalletSpendAction` info via a golang/JS bridge

The go code that handles `MsgWalletSpendAction` sends it over a golang/JS bridge. `startWalletFactory.js` registers a handler that relays such messages to the walletFactory contract (after starting the smart wallet contract).

Likewise, `MsgProvision` is used by the `walletFactory.js` contract to provision a smart wallet for each cosmos account. Wallet actions are sent to the particular smart wallet object for that owner account.

Then the smart wallet unpacks the offerSpec, uses it to conjure up an invitation in one of several ways, calls `E(zoe).offer(...)`. It also _very carefully_ extracts payments from purses to put into offers and likewise returns payouts to those purses.

The contract uses `chainStorage` to publish various kinds of state to make all this observable by clients over RPC.

**Comment by dckc** [comment, marked as answer]
a "big picture" of smart-wallet style apps is now in the docs: [Smart Wallet Dapp Architecture](https://docs.agoric.com/guides/getting-started/contract-rpc.html)

```mermaid
sequenceDiagram
  actor client

  box aqua Cosmos SDK layer
  participant chain
  end

  box darksalmon Hardened JS VM
  participant walletFactory
  participant zoe
  participant contract
  end

    client->>chain: Offer tx
    chain->>walletFactory: Offer Spec
    walletFactory->>+zoe: Offer
    walletFactory-->>chain: Offer status
    client->>+chain: vstorage query
    chain-->>-client: Offer status
    zoe->>+contract: Proposal
    contract-->>-zoe: reallocations
    zoe-->>-walletFactory: payouts
    walletFactory-->>chain: Offer status
    client->>+chain: vstorage query
    chain-->>-client: Offer status
```

_while drafting it, I (ab)used github's nifty support for hosting images in comments..._

<details><summary>diagram drafts</summary>

screenshot:
![image](https://github.com/Agoric/agoric-sdk/assets/150986/5d351ad8-76c0-4888-ab89-8eaa698b96b8)


successive elaboration:

```mermaid
sequenceDiagram
box Zoe API (simplified)

  actor client
  participant zoe
  participant contract
end

    client->>+zoe: offer(invitation, {give, want}, pmts)
    zoe->>+contract: {give, want}
    contract-->>-zoe: reallocate()
    zoe-->>-client: payout pmts

```

![image](https://github.com/Agoric/agoric-sdk/assets/150986/94aa1e36-e7e7-4280-a5a3-2f9ab092c6a2)



```mermaid
sequenceDiagram
box Zoe API (simplified)

  actor client
  participant zoe
  participant contract
end

    client->>contract: makeTradeInvitation()
    contract-->>zoe: makeInvitation('trade')
    zoe-->>client: invitation
```

submitting offers in cosmos Protobuf messages:

```mermaid
sequenceDiagram
  actor client

  box aqua Cosmos SDK layer
  participant chain
  end

  box darksalmon Hardened JS VM
  participant walletFactory
  participant zoe
  participant contract
  end

    client->>chain: Offer tx
    chain->>walletFactory: Offer Spec
    walletFactory->>zoe: Offer
    zoe->>+contract: Proposal
    contract-->>-zoe: reallocations
    zoe-->>walletFactory: payouts
```

aux info in vstorage:

```mermaid
sequenceDiagram
  actor client

  box aqua Cosmos SDK layer
  participant chain
  end

  box darksalmon Hardened JS VM
  participant contract
  end

    contract-->>chain: aux info
    client->>chain: vstorage query
    chain-->>client: aux info
```

invitationSpec

```mermaid
sequenceDiagram

  box darksalmon Hardened JS VM
  participant walletFactory
  participant zoe
  participant contract
  end

    walletFactory->>contract: makeBattleInvitation('troll')
    contract-->>zoe: makeInvitation(...)
    zoe-->>walletFactory: invitation
```

screenshot

![image](https://github.com/Agoric/agoric-sdk/assets/150986/f07b440f-0cc4-47c8-8277-f6433c84da39)

vstorage screenshot:

![image](https://github.com/Agoric/agoric-sdk/assets/150986/40cd22f0-fa01-43a9-b92a-4f0f4813a4f6)
</details>

**Comment by dckc** [comment]
This just in... @samsiegart made a [simple smart wallet app](https://github.com/samsiegart/simple-game) (context: https://github.com/Agoric/ui-kit/pull/37 ).
